### ISSUE TITLE
Xtend Batch Compiler prefers Jar Classes over Local Classes (Java Source)

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-xtend#750]
Xtend Batch Compiler prefers Jar Classes over Local Classes (Java Source)
if a local class shadows a class in a jar on the classpath

sample main call

```
package org.eclipse.xtend.core.compiler.batch;
public class Mainr {

	public static void main(String[] args) {
		String sampleDir = "/home/dietrich/Downloads/BatchCompilerIssue";
		String eclipsePluginsDir = "/home/dietrich/Downloads/Eclipse-2019-03-RC1/eclipse/plugins";
		Main.main(new String[] { "-cp",
				sampleDir + "/DullyGen/dully.jar:"+eclipsePluginsDir+"/com.google.guava_21.0.0.v20170206-1425.jar:"+eclipsePluginsDir+"/org.eclipse.xtext.xbase.lib_2.17.0.v20190304-0518.jar:"+eclipsePluginsDir+"/org.eclipse.xtend.lib_2.17.0.v20190304-0518.jar:"+eclipsePluginsDir+"/org.eclipse.xtend.lib.macro_2.17.0.v20190304-0518.jar",
				"-d", sampleDir + "/DullyGen/xtend-gen", "-td",
				sampleDir + "/DullyGen/tmp", sampleDir + "/Dully/src" });
	}

}
```

Sample projects
[BatchCompilerIssue.zip](https://github.com/eclipse/xtext-xtend/files/2970813/BatchCompilerIssue.zip)

cc @iloveeclipse @szarnekow

**Comment by cdietrich** [comment]
created a reproducing unit test

**Comment by cdietrich** [comment]
@holgerschill https://github.com/eclipse/xtext-xtend/blob/34e55150b93378927b97f257ac4ecaba3a293a57/org.eclipse.xtend.core/src/org/eclipse/xtend/core/compiler/batch/XtendBatchCompiler.java#L744 looks suspicious to me.
@szarnekow what do you think?

**Comment by szarnekow** [comment]
Could be: the tmpClassDirectory should have a higher precedence than the existing things on the CP to solve this bug.

**Comment by cdietrich** [comment]
@szarnekow but isnt the parent classloader a problem too then? it will win over the child in urlclassloader? wont it?

**Comment by szarnekow** [comment]
I think this piece of code here takes care of the parent / child precedence:
https://github.com/eclipse/xtext-xtend/blob/34e55150b93378927b97f257ac4ecaba3a293a57/org.eclipse.xtend.core/src/org/eclipse/xtend/core/compiler/batch/XtendBatchCompiler.java#L758

The used parent classloader contains usually only the JRE and the entire additional classpath is configured on the child.

**Comment by cdietrich** [comment]
in that case it is ok for the parent to win. and useparent is a deliberate choice

**Comment by holgerschill** [comment]
Fixed in 2.18
